packages/chatbot-server-mongodb-public/src/tools/fetchPage.ts

```diff
@@ -0,0 +1,140 @@
+import { z } from "zod";
+import { FindContentFunc, MongoDbPageStore, Reference } from "mongodb-rag-core";
+import { MakeReferenceLinksFunc } from "mongodb-chatbot-server";
+import { normalizeUrl } from "mongodb-rag-core/dataSources";
+import { wrapTraced } from "mongodb-rag-core/braintrust";
+import { Tool, tool, ToolExecutionOptions } from "mongodb-rag-core/aiSdk";
+
+export const FETCH_PAGE_TOOL_NAME = "fetch_page";
+export const SEARCH_ALL_FALLBACK_TEXT = "{fallback_to_search}";
+const SEARCH_ON_PAGE_LENGTH_CUTOFF = 150000;
+
+export const MongoDbFetchPageToolArgsSchema = z.object({
+  pageUrl: z.string().describe("URL of page to fetch"),
+  query: z.string().describe("Search query"),
+});
+
+export type MongoDbFetchPageToolArgs = z.infer<
+  typeof MongoDbFetchPageToolArgsSchema
+>;
+
+export type FetchPageToolResult = {
+  text: string;
+  references?: Reference[];
+};
+
+export type FetchPageTool = Tool<
+  typeof MongoDbFetchPageToolArgsSchema,
+  FetchPageToolResult
+> & {
+  execute: (
+    args: MongoDbFetchPageToolArgs,
+    options: ToolExecutionOptions
+  ) => PromiseLike<FetchPageToolResult>;
+};
+
+export interface MakeFetchPageToolParams {
+  loadPage: MongoDbPageStore["loadPage"];
+  findContent: FindContentFunc;
+  makeReferences: MakeReferenceLinksFunc;
+  searchFallbackText?: string;
+  pageLengthCutoff?: number;
+}
+
+export function makeFetchPageTool({
+  loadPage,
+  findContent,
+  makeReferences,
+  searchFallbackText = SEARCH_ALL_FALLBACK_TEXT,
+  pageLengthCutoff = SEARCH_ON_PAGE_LENGTH_CUTOFF,
+}: MakeFetchPageToolParams): FetchPageTool {
+  return tool({
+    parameters: MongoDbFetchPageToolArgsSchema,
+    description: "Fetch all content for a specific URL",
+    experimental_toToolResultContent(result) {
+      return [{ type: "text", text: result.text }];
+    },
+    execute: wrapTraced(
+      async function (
+        args: MongoDbFetchPageToolArgs,
+        _options: ToolExecutionOptions
+      ): Promise<FetchPageToolResult> {
+        const normalizedUrl = normalizeUrl({ url: args.pageUrl });
```

Comment:
Can we add a couple `logRequest` calls to this tool (and maybe the search tool as well)? I know that we log to Braintrust here but it would be nice to also see that we're invoking a tool and then that it succeeded in our server logs.

Don't need to log page content but maybe the fetched URL for this tool + a list of URLs for the search tool.